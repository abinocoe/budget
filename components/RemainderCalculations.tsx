import React from "react";
import { StyleSheet, Text, View } from "react-native";

interface Props {
  dailyAllowance: number;
  daysElapsed: number;
  daysRemaining: number;
  totalAllowance: number;
  totalSpent: number;
}

const RemainderCalculations = ({
  dailyAllowance,
  daysElapsed,
  daysRemaining,
  totalAllowance,
  totalSpent
}: Props) => {
  const amountRemaining = totalAllowance - totalSpent;
  const decimalise = (amount: number) => (amount / 100).toFixed(2).toString();
  return (
    <View>
      <View style={styles.row}>
        <Text>{`Total Remaining: ${decimalise(amountRemaining)}`}</Text>
      </View>
      <View style={styles.row}>
        <Text>
          {`Amount credit: ${decimalise(
            dailyAllowance * daysElapsed - totalSpent
          )}`}
        </Text>
      </View>
      <View style={styles.row}>
        <Text>
          {`Average per day remain: ${decimalise(
            amountRemaining / daysRemaining
          )}`}
        </Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  row: {
    flexDirection: "row"
  }
});

export default RemainderCalculations;
