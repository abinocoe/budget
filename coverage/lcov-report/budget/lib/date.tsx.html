<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for budget/lib/date.tsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">budget/lib</a> date.tsx
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">80% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>12/15</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.67% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>5/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>3/3</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">80% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>12/15</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38</td><td class="line-coverage quiet"><span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import _ from "lodash";
&nbsp;
const today = new Date();
const dateToday = today.getDate();
&nbsp;
export const numberOfDaysInStartMonth = (startMonth: number) =&gt;
  32 -
  new Date(
    startMonth === 0 ? today.getFullYear() - 1 : <span class="branch-1 cbranch-no" title="branch not covered" >today.getFullYear(),</span>
    startMonth,
    32
  ).getDate();
&nbsp;
export const getDisplayDays = (startDate: number, startMonth: number) =&gt; {
  let days: number[];
  <span class="missing-if-branch" title="else path not taken" >E</span>if (dateToday &lt; startDate) {
    days = [
      ..._.range(startDate, numberOfDaysInStartMonth(startMonth) + 1).concat(
        dateToday === 1 ? <span class="branch-0 cbranch-no" title="branch not covered" >[1] </span>: _.range(1, dateToday + 1)
      )
    ];
  } else <span class="cstat-no" title="statement not covered" >if (dateToday === startDate) {</span>
<span class="cstat-no" title="statement not covered" >    days = [startDate];</span>
  } else {
<span class="cstat-no" title="statement not covered" >    days = _.range(startDate, dateToday + 1);</span>
  }
  return days;
};
&nbsp;
export const getPeriodStartMonth = (startDate: number) =&gt; {
  const thisMonth = today.getMonth();
  return dateToday &gt;= startDate
    ? <span class="branch-0 cbranch-no" title="branch not covered" >thisMonth</span>
    : thisMonth === 0
    ? <span class="branch-0 cbranch-no" title="branch not covered" >11</span>
    : thisMonth - 1;
};
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 13 2019 11:12:04 GMT+0000 (Greenwich Mean Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
