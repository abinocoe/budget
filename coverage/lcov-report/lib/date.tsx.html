<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/date.tsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> / <a href="index.html">lib</a> date.tsx
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">40% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>6/15</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/3</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">40% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>6/15</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38</td><td class="line-coverage quiet"><span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import _ from "lodash";
&nbsp;
const today = new Date();
const dateToday = today.getDate();
&nbsp;
export const numberOfDaysInStartMonth = <span class="fstat-no" title="function not covered" >(</span>startMonth: number) =&gt;
<span class="cstat-no" title="statement not covered" >  32 -</span>
  new Date(
    startMonth === 0 ? today.getFullYear() - 1 : today.getFullYear(),
    startMonth,
    32
  ).getDate();
&nbsp;
export const getDisplayDays = <span class="fstat-no" title="function not covered" >(</span>startDate: number, startMonth: number) =&gt; {
  let days: number[];
<span class="cstat-no" title="statement not covered" >  if (dateToday &lt; startDate) {</span>
<span class="cstat-no" title="statement not covered" >    days = [</span>
      ..._.range(startDate, numberOfDaysInStartMonth(startMonth) + 1).concat(
        dateToday === 1 ? [1] : _.range(1, dateToday + 1)
      )
    ];
  } else <span class="cstat-no" title="statement not covered" >if (dateToday === startDate) {</span>
<span class="cstat-no" title="statement not covered" >    days = [startDate];</span>
  } else {
<span class="cstat-no" title="statement not covered" >    days = _.range(startDate, dateToday + 1);</span>
  }
<span class="cstat-no" title="statement not covered" >  return days;</span>
};
&nbsp;
export const getPeriodStartMonth = <span class="fstat-no" title="function not covered" >(</span>startDate: number) =&gt; {
  const thisMonth = <span class="cstat-no" title="statement not covered" >today.getMonth();</span>
<span class="cstat-no" title="statement not covered" >  return dateToday &gt;= startDate</span>
    ? thisMonth
    : thisMonth === 0
    ? 11
    : thisMonth - 1;
};
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Feb 11 2019 10:11:00 GMT+0000 (Greenwich Mean Time)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
</body>
</html>
