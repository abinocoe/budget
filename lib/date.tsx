import _ from "lodash";

export const today = new Date();
export const dateToday = today.getDate();

const dayStrings = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];

export const numberOfDaysInStartMonth = (startMonth: number) =>
  32 -
  new Date(
    startMonth === 0 ? today.getFullYear() - 1 : today.getFullYear(),
    startMonth,
    32
  ).getDate();

export const getDisplayDays = (startDate: number, startMonth: number) => {
  let days: number[];
  if (dateToday < startDate) {
    days = [
      ..._.range(startDate, numberOfDaysInStartMonth(startMonth) + 1).concat(
        dateToday === 1 ? [1] : _.range(1, dateToday + 1)
      )
    ];
  } else if (dateToday === startDate) {
    days = [startDate];
  } else {
    days = _.range(startDate, dateToday + 1);
  }
  return days;
};

export const getPeriodStartMonth = (startDate: number) => {
  const thisMonth = today.getMonth();
  return dateToday >= startDate
    ? thisMonth
    : thisMonth === 0
    ? 11
    : thisMonth - 1;
};

export const getFriendlyDate = (day: number) => {
  const mutableDate = new Date();
  if (day > dateToday) {
    mutableDate.setDate(day);
    mutableDate.setMonth(today.getMonth() - 1);
  } else {
    mutableDate.setDate(day);
  }
  return `${dayStrings[mutableDate.getDay()]} ${day}${getOrdinal(day)}`;
};

const getOrdinal = (day: number) => {
  if (day > 3 && day < 21) return "th";
  switch (day % 10) {
    case 1:
      return "st";
    case 2:
      return "nd";
    case 3:
      return "rd";
    default:
      return "th";
  }
};
