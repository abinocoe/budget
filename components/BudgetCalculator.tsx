import React, { Component } from "react";
import { AsyncStorage } from "react-native";
import SafeAreaView from "react-native-safe-area-view";

import {
  getDisplayDays,
  getPeriodStartMonth,
  numberOfDaysInStartMonth
} from "../lib/date";

import DateSelection from "./DateSelection";
import DayInputs from "./DayInputsContainer";
import PeriodAllowanceInput from "./PeriodAllowance";
import RemainderCalculations from "./RemainderCalculations";

interface State {
  dailyAmountsSpent: any;
  periodTotalAmount: number;
  periodTotalSpent: number;
  periodStartDate: number;
}

class BudgetCalculator extends Component<{}, State> {
  public constructor(props: any) {
    super(props);
    this.state = {
      dailyAmountsSpent: {},
      periodTotalAmount: 50000,
      periodTotalSpent: 0,
      periodStartDate: 25
    };
    this.updatePeriodTotal.bind(this);
    this.updateDayAmountSpent.bind(this);
    this.updatePeriodStart.bind(this);
  }

  public componentWillMount() {
    this.fetchData();
    this.handlePreviousMonths();
  }

  public updatePeriodTotal = (newTotal: string) => {
    this.setState(
      { periodTotalAmount: parseInt(newTotal, 10) },
      this.storeData
    );
  };

  public updatePeriodStart = (newStartDate: number) => {
    this.setState({ periodStartDate: newStartDate }, this.storeData);
  };

  public updateDayAmountSpent = (amountSpent: number, date: number) => {
    let periodTotalSpent = 0;
    const dailyAmountsSpent = {
      ...this.state.dailyAmountsSpent,
      [date]: amountSpent
    };
    if (isNaN(amountSpent)) {
      delete dailyAmountsSpent[date];
    }
    const dayTotalArray = Object.keys(dailyAmountsSpent);
    if (dayTotalArray.length > 0) {
      periodTotalSpent = dayTotalArray
        .map((k: string) => dailyAmountsSpent[k])
        .reduce((a, b) => a + b);
    }
    this.setState(
      {
        dailyAmountsSpent,
        periodTotalSpent
      },
      this.storeData
    );
  };

  public render() {
    const periodStartMonth = getPeriodStartMonth(this.state.periodStartDate);
    const daysInStartMonth = numberOfDaysInStartMonth(periodStartMonth);
    const daysElapsed = getDisplayDays(
      this.state.periodStartDate,
      periodStartMonth
    ).length;
    const daysRemaining = daysInStartMonth - daysElapsed;
    const dailyAllowance = Math.floor(
      this.state.periodTotalAmount / daysInStartMonth
    );
    return (
      <SafeAreaView
        forceInset={{ top: "always", bottom: "always" }}
        style={{ flex: 1 }}
      >
        <DateSelection
          intervalStartDate={this.state.periodStartDate}
          updateIntervalStartDate={this.updatePeriodStart}
        />
        <PeriodAllowanceInput
          allowanceAmount={this.state.periodTotalAmount}
          updateAllowance={this.updatePeriodTotal}
        />
        <RemainderCalculations
          dailyAllowance={dailyAllowance}
          daysElapsed={daysElapsed}
          daysRemaining={daysRemaining}
          totalAllowance={this.state.periodTotalAmount}
          totalSpent={this.state.periodTotalSpent}
        />
        <DayInputs
          startNumber={this.state.periodStartDate}
          updateAmountSpent={this.updateDayAmountSpent}
          amountsSpent={this.state.dailyAmountsSpent}
        />
      </SafeAreaView>
    );
  }

  private handlePreviousMonths = async () => {
    const newDate = new Date();
    newDate.setMonth(newDate.getMonth(), -1);
    const lastMonthsData = await AsyncStorage.getItem(`${newDate.getMonth()}`);
    if (lastMonthsData !== null) {
      const data = JSON.parse(lastMonthsData);
      if (data.periodTotalAmount && data.periodTotalSpent) {
        await AsyncStorage.setItem(
          `${newDate.getMonth()}`,
          (data.periodTotalAmount - data.periodTotalSpent).toString()
        );
      }
    }
    newDate.setMonth(newDate.getMonth(), -1);
    const monthBeforeThat = await AsyncStorage.getItem(`${newDate.getMonth()}`);
    if (monthBeforeThat !== null) {
      await AsyncStorage.removeItem(`${newDate.getMonth()}`);
    }
  };

  private storeData = async () => {
    const today = new Date();
    const data = this.state;
    const jsonOfItem = await AsyncStorage.setItem(
      `${today.getMonth()}`,
      JSON.stringify(data)
    );
    return jsonOfItem;
  };

  private fetchData = async () => {
    const today = new Date();
    const savedData = await AsyncStorage.getItem(`${today.getMonth()}`);
    if (savedData !== null) {
      const data = JSON.parse(savedData);
      this.setState({
        dailyAmountsSpent: data.dailyAmountsSpent,
        periodTotalAmount: data.periodTotalAmount,
        periodTotalSpent: data.periodTotalSpent,
        periodStartDate: data.periodStartDate
      });
    }
  };
}

export default BudgetCalculator;
