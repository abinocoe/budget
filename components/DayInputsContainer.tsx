import _ from "lodash";
import React from "react";
import { ScrollView, View } from "react-native";

import { getDisplayDays, getPeriodStartMonth } from "../lib/date";

import Day from "./Day";

interface Props {
  amountsSpentThisMonth: { [index: string]: number };
  amountsSpentLastMonth: { [index: string]: number };
  startNumber: number;
  updateAmountSpent: (amount: number, day: number) => void;
}

const DayInputsContainer = ({
  amountsSpentThisMonth,
  amountsSpentLastMonth,
  startNumber,
  updateAmountSpent
}: Props) => {
  const days = getDisplayDays(
    startNumber,
    getPeriodStartMonth(startNumber)
  ).reverse();
  return (
    <View style={{ flex: 1 }}>
      <ScrollView style={{ flex: 1, marginVertical: 30 }}>
        {days.map(date => (
          <Day
            amountSpent={
              amountsSpentThisMonth[date] || amountsSpentLastMonth[date]
            }
            date={date}
            key={date}
            updateAmountSpent={updateAmountSpent}
          />
        ))}
      </ScrollView>
    </View>
  );
};

export default DayInputsContainer;
