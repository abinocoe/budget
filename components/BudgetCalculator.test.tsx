import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import { AsyncStorage } from "react-native";

import { today } from "../lib/date";

import BudgetCalculator from "./BudgetCalculator";
import DayInputsContainer from "./DayInputsContainer";
import RemainderCalculations from "./RemainderCalculations";

const currentDate = new Date();
const nextTick = () => new Promise(resolve => setTimeout(resolve, 0));

beforeAll(() => {
  jest.mock("AsyncStorage");
});

beforeEach(() => {
  jest.clearAllMocks();
});

afterEach(() => {
  today.setDate(currentDate.getDate());
  today.setMonth(currentDate.getMonth());
  today.setFullYear(currentDate.getFullYear());
});

it("renders correctly", async () => {
  today.setDate(25);
  today.setMonth(2);
  today.setFullYear(2019);
  await nextTick();
  const output = shallow(<BudgetCalculator />);
  expect(toJson(output)).toMatchSnapshot();
});

it("retrieves data on mount", async () => {
  shallow(<BudgetCalculator />);
  await nextTick();
  expect(AsyncStorage.multiGet).toHaveBeenCalledTimes(1);
  expect(AsyncStorage.getItem).toHaveBeenCalledTimes(2);
});

it("recalculates total amount spent when state updates", async () => {
  jest
    .spyOn(AsyncStorage, "getItem")
    .mockImplementationOnce(() => Promise.resolve("50000"))
    .mockImplementationOnce(() => Promise.resolve("25"));
  const multiGetMock = jest.spyOn(AsyncStorage, "multiGet");
  multiGetMock.mockImplementationOnce(() =>
    Promise.resolve([
      ["2-2019", '{ "1": 1000, "2": 1000 }'],
      ["1-2019", '{ "31": 1000, "30": 1000 }']
    ] as [string, string][])
  );
  const output = shallow(<BudgetCalculator />);
  expect(output.state("periodTotalSpent")).toEqual(0);
  await nextTick();
  expect(output.state("periodTotalSpent")).toEqual(4000);
});

describe("updatePeriodTotal", () => {
  it("handles period allowance update from child", async () => {
    const output = shallow(<BudgetCalculator />);
    await nextTick();
    output.find(RemainderCalculations).prop("updateAllowance")(40000);
    expect(output.state("periodTotalAmount")).toEqual(40000);
  });
});

describe("updatePeriodStart", () => {
  it("handles period start date update from child", async () => {
    const output = shallow(<BudgetCalculator />);
    await nextTick();
    expect(output.state("periodStartDate")).toEqual(1);
    output.find(RemainderCalculations).prop("updateIntervalStartDate")(5);
    expect(output.state("periodStartDate")).toEqual(5);
  });
});

describe("updateDayAmountSpent", () => {
  it("handles amount spent update for a specific day from a child", async () => {
    today.setDate(3);
    jest
      .spyOn(AsyncStorage, "getItem")
      .mockImplementationOnce(() => Promise.resolve("50000"))
      .mockImplementationOnce(() => Promise.resolve("25"));
    const multiGetMock = jest.spyOn(AsyncStorage, "multiGet");
    multiGetMock.mockImplementationOnce(() =>
      Promise.resolve([
        ["2-2019", '{ "1": 1000, "2": 1000 }'],
        ["1-2019", '{ "31": 1000, "30": 1000 }']
      ] as [string, string][])
    );
    const output = shallow(<BudgetCalculator />);
    await nextTick();
    output.find(DayInputsContainer).prop("updateAmountSpent")(500, 31);
    expect(output.state("lastMonthAmountsSpent")).toEqual({
      31: 500,
      30: 1000
    });
  });
});
