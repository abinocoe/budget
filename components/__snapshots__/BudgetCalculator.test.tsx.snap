// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BudgetCalculator />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <RemainderCalculations
<<<<<<< HEAD
          dailyAllowance={1613}
=======
          dailyAllowance={1612}
>>>>>>> master
          daysElapsed={1}
          daysRemaining={31}
          intervalStartDate={25}
          totalAllowance={50000}
          totalSpent={0}
          updateAllowance={[Function]}
          updateIntervalStartDate={[Function]}
        />,
        <DayInputsContainer
          amountsSpent={Object {}}
          startNumber={25}
          updateAmountSpent={[Function]}
        />,
      ],
      "forceInset": Object {
        "bottom": "always",
        "top": "always",
      },
      "style": Object {
        "flex": 1,
        "marginHorizontal": "5%",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
<<<<<<< HEAD
          "dailyAllowance": 1613,
=======
          "dailyAllowance": 1612,
>>>>>>> master
          "daysElapsed": 1,
          "daysRemaining": 31,
          "intervalStartDate": 25,
          "totalAllowance": 50000,
          "totalSpent": 0,
          "updateAllowance": [Function],
          "updateIntervalStartDate": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "amountsSpent": Object {},
          "startNumber": 25,
          "updateAmountSpent": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <RemainderCalculations
<<<<<<< HEAD
            dailyAllowance={1613}
=======
            dailyAllowance={1612}
>>>>>>> master
            daysElapsed={1}
            daysRemaining={31}
            intervalStartDate={25}
            totalAllowance={50000}
            totalSpent={0}
            updateAllowance={[Function]}
            updateIntervalStartDate={[Function]}
          />,
          <DayInputsContainer
            amountsSpent={Object {}}
            startNumber={25}
            updateAmountSpent={[Function]}
          />,
        ],
        "forceInset": Object {
          "bottom": "always",
          "top": "always",
        },
        "style": Object {
          "flex": 1,
          "marginHorizontal": "5%",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "dailyAllowance": 1613,
            "daysElapsed": 1,
            "daysRemaining": 31,
            "intervalStartDate": 25,
            "totalAllowance": 50000,
            "totalSpent": 0,
            "updateAllowance": [Function],
            "updateIntervalStartDate": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "amountsSpent": Object {},
            "startNumber": 25,
            "updateAmountSpent": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
